### PR TITLE
REGR: fix eval with inplace=True to correctly update column values inplace

### LINKED ISSUE #47449: REGR: DataFrame.eval not respecting inplace argument in pandas 1.4
```python
import pandas as pd
import numpy as np

# Example for eval
pdf = pd.DataFrame({"A": range(1, 6), "B": range(10, 0, -2), "C": range(11, 16)})
pser = pdf.A
pdf.eval("A = B + C", inplace=True)
# pser != pdf.A after 1.4
```

_Originally posted by @Yikun in https://github.com/pandas-dev/pandas/issues/47188#issuecomment-1148168713_

### PR DESCRIPTION
**Pull request by CloseChoice**
- [x] closes #47449
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

The bad commit that causes this issue (see https://github.com/pandas-dev/pandas/pull/43406) is the rootcause for multiple others, so am not sure if we want to tackle the overall problem bit by bit or if there is a generic solution to all of them.

### CODE REVIEW

File: pandas/core/computation/eval.py [hunk 1]
```diff
@@ -384,7 +386,15 @@ def eval(
...
                 with warnings.catch_warnings(record=True):
                     # TODO: Filter the warnings we actually care about here.
-                    target[assigner] = ret
+                    if (
+                        inplace is True
+                        and hasattr(target, "columns")
+                        and assigner in target.columns
+                    ):
+                        loc = target.columns.get_loc(assigner)
+                        target._iset_item_mgr(loc, np.array(ret), inplace=inplace)
```
[the remark this answers is not in the document]
Comment (CloseChoice) [reply pandas/core/computation/eval.py]: Thanks, that helps a lot.

**Comment by CloseChoice** [comment]
> Could you add a whatsnew? Is there a suitable class for this test?
> 
> The root cause was intended, so we have to fix them where they occur.

Thanks for the answer. I added a whatsnew entry.

File: pandas/core/computation/eval.py [hunk 2]
```diff
@@ -384,7 +384,10 @@ def eval(
             try:
                 with warnings.catch_warnings(record=True):
                     # TODO: Filter the warnings we actually care about here.
-                    target[assigner] = ret
+                    if hasattr(target, "loc"):
```
Comment (phofl) [inline pandas/core/computation/eval.py:387]: This is not ideal. What could target be here? Dict, series, DataFrame, something else?

File: doc/source/whatsnew/v1.4.4.rst [hunk 3]
```diff
@@ -14,8 +14,7 @@ including other versions of pandas.
 
 Fixed regressions
 ~~~~~~~~~~~~~~~~~
--
--
+- Fixed regression in :meth:`DataFrame.eval` called with ``inplace=True`` had no effect (:issue:`47449`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.4.4.rst:17]: the `inplace` argument also affects the return type so can perhaps reword for clarity and to be consistent with the release note for #47327

```suggestion
- Fixed regression in :meth:`DataFrame.eval` creating a copy when updating inplace (:issue:`47449`)
```

File: pandas/tests/computation/test_eval.py [hunk 4]
```diff
@@ -1890,6 +1890,14 @@ def test_negate_lt_eq_le(engine, parser):
         tm.assert_frame_equal(result, expected)
 
 
+def test_set_inplace():
+    # GH 47449
+    df = DataFrame({"A": range(1, 6), "B": range(10, 0, -2), "C": range(11, 16)})
+    expected = Series([21, 20, 19, 18, 17], index=[0, 1, 2, 3, 4], name="A")
+    df.eval("A = B + C", inplace=True)
+    tm.assert_series_equal(df["A"], expected)
```
Comment (simonjayhawkins) [inline pandas/tests/computation/test_eval.py:1898]: we need to test that a view on the DataFrame is updated. Can you replicate the code sample in the issue OP. i.e. assign a column to a variable (creating a view) before the `eval` method call. (similar to tests added in #47327)

Comment (CloseChoice) [reply pandas/tests/computation/test_eval.py:1899]: done

File: doc/source/whatsnew/v1.4.4.rst (hunk 3 above)
Comment (CloseChoice) [reply doc/source/whatsnew/v1.4.4.rst]: done

**Comment by jorisvandenbossche** [comment]
@CloseChoice I started converting the suggestion at https://github.com/pandas-dev/pandas/issues/47449#issuecomment-1180701954 into a PR, missing that you already opened one with a very similar fix! I hope you don't mind, but so I merged some of the changes I started doing in your PR:

- Also checking for `inplace` in eval.py (when not doing inplace, I think the `target[assigner] = ret` without loc is faster), as suggested in https://github.com/pandas-dev/pandas/issues/47449#issuecomment-1180701954). In addition I also replaced (Series, DataFrame) with NDFrame
- I slightly updated the test (used some simpler (less) values), and also added a check for a viewing Series to be updated (in addition to viewing DataFrame)

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by phofl** [comment]
Is it possible to get there with a MultiIndex? Similar to https://github.com/pandas-dev/pandas/pull/47774 where the dimension reduction bothers us

**Comment by jorisvandenbossche** [comment]
I am not sure if `eval` supports referring (or assigning) to MultiIndex columns (I also don't directly see a test in `test_eval.py` that involved MultiIndex). I tried a few things that all fail:

```
# those two give syntax error when parsing the ast
df.eval("('A', 'a') = ('A', 'b') + ('B', 'a')")
df.eval("\"('A', 'a')\" = \"('A', 'b')\" + \"('B', 'a')\"")
# those two each give another error
df.eval("D = ('A', 'b') + ('B', 'a')") 
df.eval("D = \"('A', 'b')\" + \"('B', 'a')\"")
```

If not using existing column names, but just assigning a new column, that still seems to work with this PR:

```
In [1]: df = DataFrame({("A", "a"): [1, 2, 3], ("A", "b"): [4, 5, 6], ("B", "a"): [7, 8, 9]})

In [2]: df
Out[2]: 
   A     B
   a  b  a
0  1  4  7
1  2  5  8
2  3  6  9

In [3]: df.eval("D = 1")
Out[3]: 
   A     B  D
   a  b  a   
0  1  4  7  1
1  2  5  8  1
2  3  6  9  1

In [4]: df.eval("E = 1", inplace=True)

In [5]: df
Out[5]: 
   A     B  E
   a  b  a   
0  1  4  7  1
1  2  5  8  1
2  3  6  9  1
```

**Comment by phofl** [comment]
Thanks for checking

**Comment by CloseChoice** [comment]
@jorisvandenbossche I don't mind at all. Thanks for picking this up and improving upon it.

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
Greenish, merging

**Comment by phofl** [comment]
thx @CloseChoice

**Pull request merged**